c10/core/DeviceType.cpp

```diff
@@ -143,11 +143,15 @@ void register_privateuse1_backend(const std::string& backend_name) {
       privateuse1_backend_name);
 
   privateuse1_backend_name = backend_name;
-  // Invariant: once this flag is set, privateuse1_backend_name is NEVER written
-  // to.
+  // Use unregister_privateuse1_backend to unregister privateuse1
   privateuse1_backend_name_set.store(true, std::memory_order_relaxed);
 }
 
+void unregister_privateuse1_backend() {
+  privateuse1_backend_name = "";
+  privateuse1_backend_name_set.store(false, std::memory_order_relaxed);
```

Comment:
This is not properly thread safe

Reply:
fixed

Reply:
This doesn't work as the lock-less read path checks this flag first. Also why "relaxed" here?

Reply:
I looked into this issue and it seems that using "release" here is more appropriate. Additionally, the "relax" in `register_privateuse1_backend` should also be changed to "release". These two changes, along with the reading, will form a Release-Acquire ordering to ensure the correct sequence.

Reply:
But if the read code runs in between the two lines here, it will read the flag as true, but find the name to be an empty string ?

Reply:
The same issue you mentioned also exists for the `register_privateuse1_backend` function. As noted in the comments above, 
```
// (2) Atomic is needed during reading:
//     Whenever a user prints a privateuse1 device name, they need to read this
//     variable. Although unlikely, we'll data race if someone else is trying to
//     set this variable at the same time that another thread is print the
//     device name. We could re-use the same mutex, but reading the atomic will
//     be much faster.
```  
the original design also accepted this kind of data race. Therefore, there are two approaches here: one is to maintain the original design and accept the risk of this data race, and the other is to add a lock during reading and no longer use atomic operations.